src/game/common/rts/science.cpp

```diff
@@ -55,16 +74,139 @@ void ScienceStore::Reset()
 
 ScienceType ScienceStore::Lookup_Science(const char *name)
 {
-    NameKeyType key = g_theNameKeyGenerator->Name_To_Key(name);
+    ScienceType key = static_cast<ScienceType>(g_theNameKeyGenerator->Name_To_Key(name));
 
-    for (auto it = m_infoVec.begin(); it != m_infoVec.end(); ++it) {
-        ScienceInfo *info = (ScienceInfo *)((*it)->Get_Final_Override());
+    captainslog_relassert(
+        Has_Science(key), 0xDEAD0006, "Science name %s not known! (Did you define it in science.ini?)", name);
+    return key;
+}
+
+// zh: 0x004887F0 wb: 0x0072727E
+ScienceType ScienceStore::Get_Science_From_Internal_Name(const Utf8String &name) const
+{
+    if (name.Is_Empty()) {
+        return ScienceType::SCIENCE_INVALID;
+    }
+    return static_cast<ScienceType>(g_theNameKeyGenerator->Name_To_Key(name.Str()));
+}
 
+// zh: 0x00488820 wb: 0x007272BE
+Utf8String ScienceStore::Get_Internal_Name_From_Science(ScienceType science) const
+{
+    if (science == ScienceType::SCIENCE_INVALID) {
+        return Utf8String::s_emptyString;
+    }
+    return g_theNameKeyGenerator->Key_To_Name(static_cast<NameKeyType>(science));
+}
+
+// wb 0x0072731A
+std::vector<Utf8String> ScienceStore::Get_All_Science() const
+{
+    std::vector<Utf8String> all_science;
+    for (const auto *science : m_infoVec) {
+        const auto key = static_cast<const ScienceInfo *>(science->Get_Final_Override())->Get_Name_Key();
+        all_science.push_back(g_theNameKeyGenerator->Key_To_Name(static_cast<NameKeyType>(key)));
+    }
+    return all_science;
+}
+
+// zh: 0x00488C40 wb: 0x00727753
+int32_t ScienceStore::Get_Science_Purchase_Cost(ScienceType science) const
+{
+    const auto *info = Get_Science_Info(science);
+    return info != nullptr ? info->Get_Purchase_Cost() : 0;
+}
+
+// zh: 0x00488C90 wb: 0x00727781
+bool ScienceStore::Is_Science_Grantable(ScienceType science) const
+{
+    const auto *info = Get_Science_Info(science);
+    return info != nullptr ? info->Is_Grantable() : false;
+}
+
+// zh: 0x00488CE0 wb: 0x007277AF
+void ScienceStore::Get_Name_And_Description(ScienceType science, Utf16String &name, Utf16String &description)
```

Comment:
Minor: Maybe call this `ScienceType type` in all these functions instead? Just `science` reads ambiguous here. 2 lines below the `ScienceInfo` is also just called `info`. So `type` seems appropriate. Or `science_type`.

Reply:
It looks like `science` was used in vanilla as the name for a lot of the instances of `ScienceType`. I'm not overly attached to the name though. I don't think `type` would be great there are already lots of `type`s.

Reply:
I've adjusted to use just science throughout. There were a few instances that still used name_key but its now at least consistent to one thing.